File: LeetCode Problems/1001 to 1100/1026. Maximum Difference Between Node and Ancestor/1026. Maximum Difference Between Node and Ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int helper(TreeNode *root, int mn, int mx) {
        if(root == nullptr)
            return mx - mn;
        mn = min(mn, root->val);
        mx = max(mx, root->val);
        return max(helper(root->left, mn, mx), helper(root->right, mn, mx));
    }
public:
    int maxAncestorDiff(TreeNode* root) {
        return helper(root, INT_MAX, 0);
    }
};

File: LeetCode Problems/1001 to 1100/1026. Maximum Difference Between Node and Ancestor/1026. Maximum Difference Between Node and Ancestor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int helper(TreeNode root, int mn, int mx) {
        if (root == null) return mx - mn;
        mn = Math.min(mn, root.val);
        mx = Math.max(mx, root.val);
        return Math.max(helper(root.left, mn, mx), helper(root.right, mn, mx));
    }
    public int maxAncestorDiff(TreeNode root) {
        return helper(root, 10000000, 0);
    }
}
